File: analyzers/python/analyzer_bridge.py
#!/usr/bin/env python3
"""Python analyzer bridge for object escape analysis.

Dynamic execution now uses direct heap/object tracing to measure memory growth
around each target invocation while preserving the bridge's existing result
packaging and report contract.
"""

import gc
import json
import sys
import inspect
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

# Add parent directories to path
BRIDGE_DIR = Path(__file__).resolve().parent
ROOT_DIR = BRIDGE_DIR.parent.parent
GRAPHENE_HA_DIR = ROOT_DIR / "graphene_ha"
TESTS_DIR = ROOT_DIR / "tests"
TESTS_PYTHON_DIR = TESTS_DIR / "python"

# Setup paths in correct order
sys.path.insert(0, str(GRAPHENE_HA_DIR))
sys.path.insert(0, str(TESTS_PYTHON_DIR))  # Add tests/python for imports
sys.path.insert(0, str(TESTS_DIR))         # Add tests for general test modules
sys.path.insert(0, str(ROOT_DIR))

from test_harness import PythonFunctionTestHarness  # type: ignore[import-not-found]
from vulnerability_detector import VulnerabilityDetector  # type: ignore[import-not-found]
import importlib
import importlib.util
import time
import tracemalloc


ESCAPE_DESTINATIONS = {
    "heap": "heap_container",
}


def parse_target(target: str) -> Tuple[str, str]:
    """Parse target format: module:function or file.py:function."""
    if ":" not in target:
        raise ValueError(
            f"Invalid target format '{target}': must be 'module:function' or 'file.py:function'"
        )
    module_part, func_name = target.rsplit(":", 1)
    if not module_part or not func_name:
        raise ValueError(
            f"Invalid target format '{target}': must include module/file and function"
        )
    return module_part, func_name


def load_function_from_target(target: str):
    module_part, func_name = parse_target(target)
    try:
        if module_part.endswith(".py"):
            if not Path(module_part).exists():
                raise FileNotFoundError(f"Source file not found: {module_part}")
            spec = importlib.util.spec_from_file_location("target_module", module_part)
            if spec is None or spec.loader is None:
                raise ImportError(f"Cannot load spec for {module_part}")
            module = importlib.util.module_from_spec(spec)
            # Register module before execution so decorators (e.g., dataclass)
            # can resolve module globals through sys.modules during import.
            sys.modules[spec.name] = module
            spec.loader.exec_module(module)
        else:
            try:
                module = importlib.import_module(module_part)
            except ModuleNotFoundError as e:
                raise ModuleNotFoundError(f"Module not found: '{module_part}' (check PYTHONPATH or imports)")
    except (FileNotFoundError, ImportError, SyntaxError) as e:
        raise ValueError(f"Failed to load module '{module_part}': {str(e)}")
    
    if not hasattr(module, func_name):
        available = [n for n in dir(module) if not n.startswith("_")]
        raise AttributeError(f"Function '{func_name}' not found in module (available: {', '.join(available[:5])}{'...' if len(available) > 5 else ''})")
    return getattr(module, func_name)


def resolve_source_file(target: str, func: Any) -> str:
    """Resolve source file path for static analysis."""
    module_part, _ = parse_target(target)
    if module_part.endswith(".py"):
        return str(Path(module_part).resolve())

    source_path = inspect.getsourcefile(func)
    if source_path:
        return str(Path(source_path).resolve())
    return ""


def _normalize_path(path_value: str) -> Optional[Path]:
    try:
        return Path(path_value).resolve()
    except Exception:
        return None


def summarize_heap_allocations(
    allocations: List[Dict[str, Any]],
    total_growth_bytes: int,
    function_name: str,
) -> str:
    """Create a compact heap-growth summary for bridge and vulnerability output."""
    if total_growth_bytes <= 0 and not allocations:
        return ""

    fragments = [f"heap_growth:{total_growth_bytes}B"]
    if function_name:
        fragments.append(f"target:{function_name}")

    for allocation in allocations[:3]:
        fragments.append(
            f"{Path(allocation['file']).name}@L{allocation['line']}+{allocation['size_diff']}B"
        )

    return "; ".join(fragments)


def convert_heap_allocations_to_protocol(
    allocations: List[Dict[str, Any]],
    source_file: str,
    function_name: str,
    total_growth_bytes: int,
    peak_bytes: int,
) -> Dict[str, Any]:
    """Convert heap-growth observations to protocol-compatible escape details."""
    details = empty_escape_details()

    for allocation in allocations:
        allocation_site = allocation["allocation_site"]
        details["escaping_references"].append(
            {
                "variable_name": function_name or "<heap>",
                "object_type": "heap_allocation",
                "allocation_site": allocation_site,
                "escaped_via": "heap",
            }
        )
        details["escape_paths"].append(
            {
                "source": function_name or "<heap>",
                "destination": ESCAPE_DESTINATIONS["heap"],
                "escape_type": "heap",
                "confidence": "high" if allocation["size_diff"] >= 1024 else "medium",
            }
        )

    details["heap_growth_bytes"] = total_growth_bytes
    details["heap_peak_bytes"] = peak_bytes
    details["source_file"] = source_file or "unknown"
    details["heap_allocations"] = allocations
    return details


def collect_heap_trace(
    before_snapshot: tracemalloc.Snapshot,
    after_snapshot: tracemalloc.Snapshot,
    source_file: str,
    function_name: str,
    limit: int = 5,
) -> Tuple[int, int, List[Dict[str, Any]]]:
    """Compare snapshots and keep the positive allocation deltas for the target source."""
    source_path = _normalize_path(source_file) if source_file else None
    growth_candidates: List[Dict[str, Any]] = []
    fallback_candidates: List[Dict[str, Any]] = []
    total_growth_bytes = 0

    for stat in after_snapshot.compare_to(before_snapshot, "lineno"):
        if stat.size_diff <= 0:
            continue

        traceback_frame = stat.traceback[0]
        frame_path = _normalize_path(traceback_frame.filename)
        candidate = {
            "file": traceback_frame.filename,
            "line": traceback_frame.lineno,
            "size_diff": int(stat.size_diff),
            "count_diff": int(stat.count_diff),
            "traceback": [f"{frame.filename}:{frame.lineno}" for frame in stat.traceback[:5]],
            "allocation_site": f"{traceback_frame.filename}:{traceback_frame.lineno}",
            "function_name": function_name,
        }
        fallback_candidates.append(candidate)

        if source_path is None or frame_path == source_path:
            growth_candidates.append(candidate)
            total_growth_bytes += int(stat.size_diff)

    if source_path is not None and not growth_candidates:
        growth_candidates = fallback_candidates[:limit]
        total_growth_bytes = sum(candidate["size_diff"] for candidate in growth_candidates)

    return total_growth_bytes, len(growth_candidates), growth_candidates[:limit]


def diagnose_bridge_error(error_msg: str):
    raw = (error_msg or "").strip()
    lower = raw.lower()

    if "timeout" in lower or "timed out" in lower or "exceeded" in lower:
        return (
            "Timeout",
            "Inspect blocking operations and missing joins/awaits before increasing timeout.",
        )
    if (
        "target resolution" in lower
        or "missing required field: 'target'" in lower
        or "target loading failed" in lower
        or "failed to load" in lower
        or "invalid target" in lower
        or "module not found" in lower
        or "not found" in lower
    ):
        return (
            "Target Resolution",
            "Verify target signature/path and ensure the target symbol exists in the selected language module.",
        )
    if (
        "protocol/input" in lower
        or "invalid json" in lower
        or "failed to parse" in lower
        or "empty input" in lower
        or "expected json" in lower
        or "json" in lower
        or "parse" in lower
        or "stdin" in lower
        or "protocol" in lower
    ):
        return (
            "Protocol/Input",
            "Validate request JSON and ensure bridge stdin/stdout protocol fields match the orchestrator contract.",
        )
    if (
        "environment" in lower
        or "permission denied" in lower
        or "not available" in lower
        or "not found in path" in lower
        or "command not found" in lower
        or "missing tools" in lower
        or "python not found" in lower
    ):
        return (
            "Environment",
            "Check runtime/toolchain installation and PATH configuration for the selected language analyzer.",
        )
    if (
        "runtime crash" in lower
        or "panic" in lower
        or "exception" in lower
        or "traceback" in lower
        or "segmentation" in lower
    ):
        return (
            "Runtime Crash",
            "Re-run with --verbose and inspect bridge stack traces for runtime exceptions.",
        )

    return (
        "Unknown",
        "Re-run with --verbose and inspect bridge stdout/stderr for additional diagnostics.",
    )


def empty_escape_details() -> dict:
    return {"escaping_references": [], "escape_paths": []}


def _error_response(language, error_msg, session_id="unknown", analysis_mode="dynamic"):
    category, suggested_action = diagnose_bridge_error(error_msg)
    return {
        "session_id": session_id,
        "language": language,
        "analyzer_version": "1.0.0",
        "analysis_mode": analysis_mode,
        "results": [
            {
                "input_data": "<bridge-startup>",
                "success": False,
                "crashed": True,
                "output": "",
                "error": f"{category}: {error_msg}",
                "execution_time_ms": 0,
                "escape_detected": False,
                "escape_details": empty_escape_details(),
            }
        ],
        "vulnerabilities": [],
        "summary": {
            "total_tests": 1,
            "successes": 0,
            "crashes": 1,
            "timeouts": 1 if category == "Timeout" else 0,
            "escapes": 0,
            "genuine_escapes": 0,
            "crash_rate": 1.0,
        },
        "error": error_msg,
        "error_category": category,
        "suggested_action": suggested_action,
    }

def analyze(request: dict) -> dict:
    session_id = request.get("session_id", "unknown")
    target = request.get("target")
    if not target:
        return _error_response("python", "Missing required field: 'target'", session_id)
    
    inputs = request.get("inputs", [])
    if not inputs:
        # Run one dynamic probe with no positional input when the CLI omits --input.
        inputs = [None]
    repeat = request.get("repeat", 1)
    timeout_seconds = request.get("timeout_seconds", 30.0)
    analysis_mode = request.get("analysis_mode", "dynamic")

    try:
        _, function_name = parse_target(target)
    except ValueError as e:
        return _error_response("python", str(e), session_id, analysis_mode)
    
    try:
        func = load_function_from_target(target)
    except (ValueError, FileNotFoundError, ModuleNotFoundError, AttributeError) as e:
        return _error_response("python", f"Target loading failed: {str(e)}", session_id, analysis_mode)
    except Exception as e:
        return _error_response("python", f"Unexpected error loading target '{target}': {type(e).__name__}: {str(e)}", session_id, analysis_mode)

    harness = PythonFunctionTestHarness(func, timeout=timeout_seconds, prefer_main_thread=True)
    all_results = []
    source_file = resolve_source_file(target, func)
    tracemalloc.start(25)

    try:
        for input_data in inputs:
            for _ in range(repeat):
                gc.collect()
                before_snapshot = tracemalloc.take_snapshot()

                start_time = time.time()
                result = harness.run_test(input_data)
                execution_time_ms = int((time.time() - start_time) * 1000)

                gc.collect()
                after_snapshot = tracemalloc.take_snapshot()
                current_bytes, peak_bytes = tracemalloc.get_traced_memory()
                total_growth_bytes, matched_allocation_count, allocations = collect_heap_trace(
                    before_snapshot,
                    after_snapshot,
                    source_file,
                    function_name,
                )
                heap_summary = summarize_heap_allocations(allocations, total_growth_bytes, function_name)
                heap_details = convert_heap_allocations_to_protocol(
                    allocations,
                    source_file,
                    function_name,
                    total_growth_bytes,
                    int(peak_bytes),
                )

                escape_detected = bool(total_growth_bytes > 0 or matched_allocation_count > 0)

                all_results.append({
                    "input_data": input_data,
                    "success": result.success,
                    "crashed": result.crashed,
                    "output": result.output,
                    "error": result.error,
                    "execution_time_ms": execution_time_ms,
                    "escape_detected": escape_detected,
                    "escape_details": heap_details if escape_detected else empty_escape_details(),
                    "heap_growth_bytes": int(total_growth_bytes),
                    "heap_current_bytes": int(current_bytes),
                    "heap_peak_bytes": int(peak_bytes),
                    "heap_summary": heap_summary,
                })
    finally:
        tracemalloc.stop()
    
    # Analyze vulnerabilities
    detector = VulnerabilityDetector()
    
    # Convert results back to TestResult-like objects for detector
    class TestResultProxy:
        def __init__(self, data):
            self.input_data = data["input_data"]
            self.success = data["success"]
            self.crashed = data["crashed"]
            self.error = data["error"]
            self.escape_detected = data["escape_detected"]
            self.escape_details = data.get("heap_summary", "")
    
    result_proxies = [TestResultProxy(r) for r in all_results]
    analysis = detector.categorize_results(result_proxies)
    vulnerabilities = [
        {
            "input": v.input,
            "vulnerability_type": v.vulnerability_type,
            "severity": v.severity,
            "description": v.error_message,
            "escape_details": next(
                (r["escape_details"] for r in all_results if r["input_data"] == v.input and r["escape_detected"]),
                empty_escape_details(),
            ),
        }
        for v in analysis["vulnerabilities"]
    ]
    return {
        "session_id": session_id,
        "language": "python",
        "analyzer_version": "1.0.0",
        "analysis_mode": analysis_mode,
        "results": all_results,
        "vulnerabilities": vulnerabilities,
        "summary": {
            "total_tests": analysis["total_tests"],
            "successes": analysis["successes"],
            "crashes": analysis["crashes"],
            "timeouts": analysis["timeouts"],
            "escapes": analysis["escapes"],
            "genuine_escapes": analysis.get("genuine_escapes", analysis["escapes"]),
            "crash_rate": analysis["crash_rate"],
        },
    }


def main():
    try:
        input_data = sys.stdin.read()
        if not input_data.strip():
            error_msg = _error_response("python", "Empty input: expected JSON request on stdin")
            print(json.dumps(error_msg), file=sys.stderr)
            sys.exit(1)
        
        try:
            request = json.loads(input_data)
        except json.JSONDecodeError as e:
            error_msg = _error_response("python", f"Invalid JSON input: {str(e)} at line {e.lineno}, column {e.colno}")
            print(json.dumps(error_msg), file=sys.stderr)
            sys.exit(1)
        
        result = analyze(request)
        print(json.dumps(result, indent=2))
        sys.exit(0 if "error" not in result else 1)
    except BrokenPipeError:
        sys.exit(0)
    except Exception as e:
        error_msg = _error_response("python", f"Bridge critical error: {type(e).__name__}: {str(e)}")
        print(json.dumps(error_msg), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: analyzers/nodejs/analyzer_bridge.js
#!/usr/bin/env node
const path = require('path');
const async_hooks = require('async_hooks');
const { analyzeFile: runStaticAnalyzer } = require('./static_analyzer');

const TRADITIONAL_ESCAPE_TYPES = new Set(['return', 'parameter', 'global', 'closure', 'heap']);
const ESCAPE_DESTINATIONS = {
    return: 'caller',
    parameter: 'callee',
    global: 'module_scope',
    closure: 'closure_scope',
    heap: 'heap_container'
};
const IGNORED_ASYNC_RESOURCE_TYPES = new Set([
    'Timeout',
    'TIMERWRAP',
    'Immediate',
    'TickObject',
    'PROMISE',
    'Microtask'
]);

class AsyncResourceTracker {
    constructor() {
        this.baselineResources = new Set();
        this.currentResources = new Map();
        this.hook = null;
    }

    start() {
        this.baselineResources.clear();
        this.currentResources.clear();
        this.hook = async_hooks.createHook({
            init: (asyncId, type) => this.currentResources.set(asyncId, {type, created: Date.now()}),
            destroy: (asyncId) => this.currentResources.delete(asyncId)
        });
        this.hook.enable();
    }

    captureBaseline() {
        this.baselineResources = new Set(this.currentResources.keys());
    }

    stop() {
        if (this.hook) this.hook.disable();
    }

    getEscapedResources() {
        const escaped = [];
        for (const [asyncId, info] of this.currentResources.entries()) {
            if (!this.baselineResources.has(asyncId) && !IGNORED_ASYNC_RESOURCE_TYPES.has(info.type)) {
                escaped.push({task_id: String(asyncId), task_type: info.type, state: 'active'});
            }
        }
        return escaped;
    }
}

function parseTargetReference(target) {
    const delimiterIndex = target.lastIndexOf(':');
    if (delimiterIndex <= 0 || delimiterIndex >= target.length - 1) {
        throw new Error(`Invalid target format '${target}': must be 'file.js:functionName' or 'module:functionName'`);
    }

    const modulePath = target.slice(0, delimiterIndex).trim();
    const functionName = target.slice(delimiterIndex + 1).trim();
    if (!modulePath || !functionName) {
        throw new Error(`Invalid target format '${target}': module path and function name are required`);
    }

    const normalizedModulePath = modulePath.replace(/\\/g, path.sep);
    const isPathTarget = /[\\/]/.test(modulePath) || /\.(js|mjs|cjs)$/i.test(modulePath);

    return {
        modulePath,
        functionName,
        normalizedModulePath,
        isPathTarget
    };
}

function resolveModuleCandidates(normalizedModulePath) {
    const workspaceRoot = path.resolve(__dirname, '..', '..');
    const candidates = [];

    if (path.isAbsolute(normalizedModulePath)) {
        candidates.push(path.normalize(normalizedModulePath));
    }
    candidates.push(path.resolve(process.cwd(), normalizedModulePath));
    candidates.push(path.resolve(workspaceRoot, normalizedModulePath));

    return [...new Set(candidates)];
}

function summarizeStaticEscapes(escapes) {
    if (!escapes.length) {
        return '';
    }

    const fragments = escapes.map((escape) => {
        const lineSuffix = Number.isInteger(escape.line) && escape.line > 0 ? `@L${escape.line}` : '';
        return `${escape.escape_type}:${escape.variable_name || '<value>'}${lineSuffix}`;
    });

    return fragments.join('; ');
}

function convertStaticEscapesToProtocol(escapes, sourcePath) {
    const details = emptyEscapeDetails();

    for (const escape of escapes) {
        const line = Number.isInteger(escape.line) ? escape.line : 0;
        const column = Number.isInteger(escape.column) ? escape.column : 0;
        const variableName = escape.variable_name || '<value>';
        const escapeType = escape.escape_type || 'unknown';
        const confidence = escape.confidence || 'medium';
        const allocationSite = sourcePath
            ? `${sourcePath}:${line}:${column}`
            : 'unknown';

        details.escaping_references.push({
            variable_name: variableName,
            object_type: 'unknown',
            allocation_site: allocationSite,
            escaped_via: escapeType
        });

        details.escape_paths.push({
            source: variableName,
            destination: ESCAPE_DESTINATIONS[escapeType] || 'unknown',
            escape_type: escapeType,
            confidence
        });
    }

    return details;
}

function runTraditionalStaticEscapeAnalysis(sourcePath, functionName) {
    if (!sourcePath || !functionName) {
        return {
            detected: false,
            escapes: [],
            details: emptyEscapeDetails(),
            summary: ''
        };
    }

    try {
        const analysis = runStaticAnalyzer(sourcePath, functionName);
        if (!analysis || !analysis.success || !Array.isArray(analysis.escapes)) {
            return {
                detected: false,
                escapes: [],
                details: emptyEscapeDetails(),
                summary: ''
            };
        }

        const escapes = analysis.escapes.filter((escape) => TRADITIONAL_ESCAPE_TYPES.has(escape.escape_type));
        if (escapes.length === 0) {
            return {
                detected: false,
                escapes: [],
                details: emptyEscapeDetails(),
                summary: ''
            };
        }

        return {
            detected: true,
            escapes,
            details: convertStaticEscapesToProtocol(escapes, sourcePath),
            summary: summarizeStaticEscapes(escapes)
        };
    } catch (_) {
        return {
            detected: false,
            escapes: [],
            details: emptyEscapeDetails(),
            summary: ''
        };
    }
}

function loadTargetFunction(target) {
    const parsedTarget = parseTargetReference(target);
    const modulePath = parsedTarget.modulePath;
    const functionName = parsedTarget.functionName;
    let loadedModule;
    let resolvedSourcePath = null;

    try {
        if (parsedTarget.isPathTarget) {
            const candidates = resolveModuleCandidates(parsedTarget.normalizedModulePath);
            let moduleLoadError = null;
            for (const candidate of candidates) {
                try {
                    loadedModule = require(candidate);
                    resolvedSourcePath = candidate;
                    moduleLoadError = null;
                    break;
                } catch (error) {
                    moduleLoadError = error;
                }
            }

            if (!loadedModule) {
                if (moduleLoadError) {
                    throw moduleLoadError;
                }
                throw new Error(`Module file '${modulePath}' could not be resolved`);
            }
        } else {
            loadedModule = require(modulePath);
            try {
                resolvedSourcePath = require.resolve(modulePath);
            } catch (_) {
                resolvedSourcePath = null;
            }
        }

        if (typeof loadedModule[functionName] !== 'function') {
            const availableFunctions = Object.keys(loadedModule).filter((key) => typeof loadedModule[key] === 'function');
            const preview = availableFunctions.slice(0, 5);
            throw new ReferenceError(`Function '${functionName}' not found in module (available: ${preview.join(', ')}${availableFunctions.length > 5 ? '...' : ''})`);
        }

        return {
            targetFunc: loadedModule[functionName],
            functionName,
            sourcePath: resolvedSourcePath
        };
    } catch (error) {
        if (error instanceof ReferenceError) throw error;
        throw new Error(`Failed to load module '${modulePath}': ${error.message}`);
    }
}

function diagnoseBridgeError(errorMsg) {
    const raw = String(errorMsg || '').trim();
    const lower = raw.toLowerCase();

    if (lower.includes('timeout') || lower.includes('timed out') || lower.includes('exceeded')) {
        return {
            category: 'Timeout',
            suggestedAction: 'Inspect blocking operations and missing joins/awaits before increasing timeout.'
        };
    }

    if (
        lower.includes('target resolution')
        || lower.includes("missing required field: 'target'")
        || lower.includes('failed to load')
        || lower.includes('invalid target')
        || lower.includes('module not found')
        || lower.includes('function') && lower.includes('not found')
    ) {
        return {
            category: 'Target Resolution',
            suggestedAction: 'Verify target signature/path and ensure the target symbol exists in the selected language module.'
        };
    }

    if (
        lower.includes('protocol/input')
        || lower.includes('invalid json')
        || lower.includes('failed to parse')
        || lower.includes('empty input')
        || lower.includes('expected json')
        || lower.includes('json')
        || lower.includes('parse')
        || lower.includes('stdin')
        || lower.includes('protocol')
    ) {
        return {
            category: 'Protocol/Input',
            suggestedAction: 'Validate request JSON and ensure bridge stdin/stdout protocol fields match the orchestrator contract.'
        };
    }

    if (
        lower.includes('environment')
        || lower.includes('permission denied')
        || lower.includes('not available')
        || lower.includes('not found in path')
        || lower.includes('command not found')
        || lower.includes('missing tools')
        || lower.includes('node.js not found')
    ) {
        return {
            category: 'Environment',
            suggestedAction: 'Check runtime/toolchain installation and PATH configuration for the selected language analyzer.'
        };
    }

    if (
        lower.includes('runtime crash')
        || lower.includes('panic')
        || lower.includes('exception')
        || lower.includes('traceback')
        || lower.includes('segmentation')
    ) {
        return {
            category: 'Runtime Crash',
            suggestedAction: 'Re-run with --verbose and inspect bridge stack traces for runtime exceptions.'
        };
    }

    return {
        category: 'Unknown',
        suggestedAction: 'Re-run with --verbose and inspect bridge stdout/stderr for additional diagnostics.'
    };
}

function emptyEscapeDetails() {
    return {
        escaping_references: [],
        escape_paths: [],
        threads: [],
        processes: [],
        async_tasks: [],
        goroutines: [],
        other: []
    };
}

function mergeEscapeDetails(primary, secondary) {
    const merged = emptyEscapeDetails();
    const first = primary || {};
    const second = secondary || {};

    const mergeArrayField = (key) => {
        merged[key] = [
            ...(Array.isArray(first[key]) ? first[key] : []),
            ...(Array.isArray(second[key]) ? second[key] : [])
        ];
    };

    mergeArrayField('escaping_references');
    mergeArrayField('escape_paths');
    mergeArrayField('threads');
    mergeArrayField('processes');
    mergeArrayField('async_tasks');
    mergeArrayField('goroutines');
    mergeArrayField('other');

    return merged;
}

async function executeTest(targetFunc, targetLabel, input, timeoutSeconds) {
    const result = {input_data: input, success: false, crashed: false, output: '', error: '', execution_time_ms: 0, escape_detected: false, escape_details: emptyEscapeDetails()};
    const tracker = new AsyncResourceTracker();
    tracker.start();
    await new Promise((resolve) => setImmediate(resolve));
    tracker.captureBaseline();

    const heapBefore = captureHeapSnapshot();

    const startTime = Date.now();
    let timeoutHandle = null;

    try {
        const returnValue = await Promise.race([
            Promise.resolve().then(() => targetFunc(input)),
            new Promise((_, reject) => {
                timeoutHandle = setTimeout(() => reject(new Error(`Function timeout after ${timeoutSeconds}s`)), timeoutSeconds * 1000);
            })
        ]);
        result.output = String(returnValue);
        result.success = true;
    } catch (error) {
        result.crashed = true;
        result.error = `${error.name}: ${error.message}`;
    } finally {
        if (timeoutHandle !== null) {
            clearTimeout(timeoutHandle);
        }
    }

    result.execution_time_ms = Date.now() - startTime;
    await new Promise(resolve => setTimeout(resolve, 100));

    const heapAfter = captureHeapSnapshot();
    const heapGrowthBytes = Math.max(0, heapAfter.heap_used_bytes - heapBefore.heap_used_bytes);
    const heapPeakBytes = Math.max(heapAfter.heap_total_bytes, heapBefore.heap_total_bytes);

    if (heapGrowthBytes > 0) {
        const label = String(targetLabel || '<anonymous>');
        result.escape_details.escaping_references.push({
            variable_name: label,
            object_type: 'heap_allocation_delta',
            allocation_site: label,
            escaped_via: 'heap'
        });

        result.escape_details.escape_paths.push({
            source: label,
            destination: 'heap_container',
            escape_type: 'heap',
            confidence: heapGrowthBytes >= 1024 ? 'high' : 'medium'
        });

        result.escape_details.other.push(`heap_growth_bytes:${heapGrowthBytes}`);
        result.escape_details.other.push(`heap_used_before_bytes:${heapBefore.heap_used_bytes}`);
        result.escape_details.other.push(`heap_used_after_bytes:${heapAfter.heap_used_bytes}`);
        result.escape_details.other.push(`heap_peak_bytes:${heapPeakBytes}`);
    }

    const escapedResources = tracker.getEscapedResources();
    result.escape_details.async_tasks = escapedResources;
    result.escape_detected = escapedResources.length > 0 || result.escape_details.escaping_references.length > 0;
    tracker.stop();
    return result;
}

function captureHeapSnapshot() {
    if (typeof global.gc === 'function') {
        try {
            global.gc();
        } catch (_) {
            // Best-effort GC; continue with current memory view.
        }
    }

    const usage = process.memoryUsage();
    return {
        heap_used_bytes: Math.max(0, usage.heapUsed || 0),
        heap_total_bytes: Math.max(0, usage.heapTotal || 0)
    };
}

function findHeapSignal(entries, prefix) {
    if (!Array.isArray(entries)) {
        return null;
    }
    return entries.find((entry) => typeof entry === 'string' && entry.startsWith(prefix)) || null;
}

function errorResponse(error, sessionId = 'unknown') {
    const errorMsg = error instanceof Error 
        ? `${error.name}: ${error.message}`
        : typeof error === 'string' ? error : 'Unknown error';
    const diagnosis = diagnoseBridgeError(errorMsg);
    return {
        session_id: sessionId,
        language: 'javascript',
        analyzer_version: '1.0.0',
        analysis_mode: 'dynamic',
        results: [{
            input_data: '<bridge-startup>',
            success: false,
            crashed: true,
            output: '',
            error: `${diagnosis.category}: ${errorMsg}`,
            execution_time_ms: 0,
            escape_detected: false,
            escape_details: emptyEscapeDetails()
        }],
        vulnerabilities: [],
        summary: {total_tests: 1, successes: 0, crashes: 1, timeouts: diagnosis.category === 'Timeout' ? 1 : 0, escapes: 0, genuine_escapes: 0, crash_rate: 1.0},
        error: `Bridge error: ${errorMsg}`,
        error_category: diagnosis.category,
        suggested_action: diagnosis.suggestedAction
    };
}

async function analyze(request) {
    const sessionId = request.session_id || request.sessionId || 'unknown';
    const analysisMode = request.analysis_mode || request.analysisMode || 'dynamic';
    const response = {session_id: sessionId, language: 'javascript', analyzer_version: '1.0.0', analysis_mode: analysisMode, results: [], vulnerabilities: [], summary: {total_tests: 0, successes: 0, crashes: 0, timeouts: 0, escapes: 0, genuine_escapes: 0, crash_rate: 0}};
    try {
        if (!request.target) throw new Error("Missing required field: 'target'");
        if (!Array.isArray(request.inputs)) throw new Error("Missing or invalid field: 'inputs' must be an array");
        
        const loadedTarget = loadTargetFunction(request.target);
        let successes = 0, crashes = 0, timeouts = 0, escapes = 0, genuineEscapes = 0;
        
        for (const input of request.inputs) {
            for (let i = 0; i < (request.repeat || 1); i++) {
                const timeoutSeconds = request.timeout_seconds || request.timeoutSeconds || 30;
                const result = await executeTest(loadedTarget.targetFunc, request.target, input, timeoutSeconds);

                response.results.push(result);
                if (result.success) successes++;
                if (result.crashed) crashes++;
                if (result.error.includes('timeout')) timeouts++;
                if (result.escape_detected) {
                    escapes++;
                    if (!result.error.includes('timeout')) genuineEscapes++;

                    const asyncCount = result.escape_details.async_tasks.length;
                    const heapSignal = findHeapSignal(result.escape_details.other, 'heap_growth_bytes:');
                    const description = heapSignal
                        ? `Node.js heap escape signal detected (${heapSignal})${asyncCount > 0 ? ` + ${asyncCount} async resource leak(s)` : ''}`
                        : `${asyncCount} async resource(s) escaped`;

                    response.vulnerabilities.push({
                        input,
                        vulnerability_type: 'object_escape',
                        severity: 'high',
                        description,
                        escape_details: result.escape_details
                    });
                }
            }
        }
        const totalTests = response.results.length;
        response.summary = {total_tests: totalTests, successes, crashes, timeouts, escapes, genuine_escapes: genuineEscapes, crash_rate: totalTests > 0 ? crashes / totalTests : 0};
    } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        const diagnosis = diagnoseBridgeError(errorMsg);
        response.results = [{
            input_data: '<bridge-startup>',
            success: false,
            crashed: true,
            output: '',
            error: `${diagnosis.category}: ${errorMsg}`,
            execution_time_ms: 0,
            escape_detected: false,
            escape_details: emptyEscapeDetails()
        }];
        response.summary.total_tests = 1;
        response.summary.crash_rate = 1.0;
        response.summary.crashes = 1;
        response.summary.successes = 0;
        response.summary.timeouts = diagnosis.category === 'Timeout' ? 1 : 0;
        response.error = errorMsg;
        response.error_category = diagnosis.category;
        response.suggested_action = diagnosis.suggestedAction;
    }
    return response;
}

async function main() {
    try {
        const chunks = [];
        for await (const chunk of process.stdin) {
            chunks.push(chunk);
        }
        const inputData = Buffer.concat(chunks).toString('utf8');
        
        if (!inputData.trim()) {
            console.error(JSON.stringify(errorResponse(new Error('Empty input: expected JSON request on stdin'))));
            process.exit(1);
        }
        
        let request;
        try {
            request = JSON.parse(inputData);
        } catch (parseError) {
            const e = parseError instanceof SyntaxError ? parseError : new Error(String(parseError));
            console.error(JSON.stringify(errorResponse(new Error(`Invalid JSON: ${e.message}`))));
            process.exit(1);
        }
        
        const response = await analyze(request);
        console.log(JSON.stringify(response, null, 2));
        process.exit(response.error ? 1 : 0);
    } catch (error) {
        console.error(JSON.stringify(errorResponse(error)));
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = { analyze, executeTest, loadTargetFunction };
